ethcore/src/evm/interpreter/memory.rs

```diff
@@ -109,6 +114,20 @@ impl Memory for Vec<u8> {
 			Memory::resize(self, size)
 		}
 	}
+
+	fn into_return_data(mut self, offset: U256, size: U256) -> ReturnData {
+		let offset = offset.low_u64() as usize;
+		let size = size.low_u64() as usize;
+		if !is_valid_range(offset, size) {
+			return ReturnData::empty()
+		}
+		if self.len() - size > MAX_RETURN_WASTE_BYTES {
+			let mem = self.drain(offset..).take(size).collect();
```

Comment:
i really doubt the optimizer is smart enough to turn this into a single malloc + memcpy, so it ends up paying both the O(n) iteration and having to allocate `self.len() + size` bytes in total.

i would suggest either using a `drain`, `truncate`, `shrink_to_fit` combo or allocating a `vec![0; size]` and using `copy_from_slice` to pay only one of those costs.